### PR TITLE
adding NCL language

### PR DESCRIPTION
**Pull request by rpavlick**
https://github.com/search?utf8=%E2%9C%93&q=extension%3Ancl+NOT+what

### CODE REVIEW

**Comment by pchaigno** [comment]
Thanks for the pull request! Do you know if [these files](https://github.com/search?p=5&q=extension%3Ancl+NOT+begin&ref=searchresults&type=Code&utf8=%E2%9C%93) are NCL too?

**Comment by rpavlick** [comment]
Those are poorly-named documentation files for NCL functions.

**Comment by Languagelearner1111** [comment]
What's better?

**Comment by rpavlick** [comment]
extension:ncl NOT "THE_URL" NOT "ISO-8859-1"

**Comment by pchaigno** [comment]
Would it be correct to recognize these files as Text?

**Comment by rpavlick** [comment]
@pchaigno Yes.

**Comment by pchaigno** [comment]
In this case, could you add `.ncl` to the Text entry in `languages.yml` too (with a sample in `samples/Text/`)?

**Comment by rpavlick** [comment]
@pchaigno I added the example to Text and updated the license in the grammar submodule.

File: .gitmodules [hunk 1]
```diff
@@ -644,3 +644,6 @@
 [submodule "vendor/grammars/smali-sublime"]
 	path = vendor/grammars/smali-sublime
 	url = https://github.com/ShaneWilton/sublime-smali
+[submodule "vendor/grammars/language-ncl"]
+	path = vendor/grammars/language-ncl
+	url = git://github.com/rpavlick/language-ncl.git
```
Comment (pchaigno) [inline .gitmodules]: Cloning the submodule fails for me in local with this URL. Could you use `git@` or `https://`?

**Comment by rpavlick** [comment]
@pchaigno I updated the grammar submodule link to https.

**Comment by pchaigno** [comment]
It's still failing locally. I don't think you can just update the `.gitmodules` file. You'll probably have to remove the submodule and add it again to be sure :/

**Comment by pchaigno** [comment]
I'll see first if it's not an issue on my side...

**Comment by rpavlick** [comment]
@pchaigno I removed the submodule and added it back with https.

**Comment by pchaigno** [comment]
I tested the detection of `.ncl` files with 2,000 samples:

 | Text | NCL |
---- | ---- | ---- |
[1,000 Text files](https://github.com/search?p=5&q=extension%3Ancl+NOT+begin&ref=searchresults&type=Code&utf8=%E2%9C%93) | 60.39% | **39.61%** |
[1,000 NCL files](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ancl+NOT+%22THE_URL%22+NOT+%22ISO-8859-1%22&type=Code&ref=searchresults) | **7.45%** | 92.55% |

The Bayesian classifier doesn't seem to be very good at distinguishing Text and NCL files :/
We could try to improve it by adding more samples or we can define a [new heuristic rule](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.rb).
Do you want me to send you the sample files?

**Comment by rpavlick** [comment]
Yes, please do.

**Comment by pchaigno** [comment]
@rpavlick In your inbox :zap:

**Comment by rpavlick** [comment]
@pchiagno So, if I manually go through these and sort out the errors, would that help?

**Comment by pchaigno** [comment]
> @pchaigno So, if I manually go through these and sort out the errors, would that help?

Not really. It's a matter of keywords so there's not much to do there except for adding new samples.
If adding a few more samples doesn't improve things, we'll see how to define a new heuristic rule.

**Comment by rpavlick** [comment]
@pchaigno I added quite a few NCL samples.

**Comment by larsbrinkhoff** [comment]
That's a bit over the top, isn't it? :smile:

**Comment by pchaigno** [comment]
@rpavlick We currently can't add too much samples because of #2117 :/

**Comment by rpavlick** [comment]
@pchaigno @larsbrinkhoff I decreased the number of added samples.

**Comment by pchaigno** [comment]
@rpavlick Could you test the detection results in local with the samples I gave you?

**Comment by rpavlick** [comment]
@pchaigno What is the command to run that test?

**Comment by larsbrinkhoff** [comment]
bundle exec rake samples # once
    bundle exec rake test

**Comment by rpavlick** [comment]
@pchaigno 

It is getting hung up on a false detection of Frege in one of the Text samples. Do you have any suggestions for implementing a heuristic?

  1) Failure:
TestHeuristcs#test_fr_by_heuristics [/Users/rpavlick/Github/linguist/test/test_heuristics.rb:122]:
Failed for /Users/rpavlick/Github/linguist/samples/Text/NhlUpdateWorkstation-help.ncl.
--- expected
+++ actual
@@ -1 +1 @@
-[#<Linguist::Language name=Text>]
+[#<Linguist::Language name=Frege>]

**Comment by pchaigno** [comment]
#2441 should fix this. In the meantime, you can change `"Text" => all_fixtures("Text")` to `"Text" => all_fixtures("Text", "*.fr")` in [`test_heuristics.rb`](https://github.com/github/linguist/blob/master/test/test_heuristics.rb#L125).

**Comment by rpavlick** [comment]
@pchaigno 

From a large sample (n=951) of known NCL files taken from Github, it is now predicting with ~98% accuracy. 

For a large sample of  (n=751) other files with extension .ncl, it is around 92%. From those, nearly all of the errors come from one github repository (https://github.com/fanghuan/ncl) and they all contain the text strings, "THE_URL:" and "THE_TITLE:".

File: Rakefile [hunk 2]
```diff
@@ -62,7 +62,7 @@ namespace :benchmark do
 
     corpus = File.expand_path(ENV["CORPUS"] || "samples")
 
-    require 'linguist/language'
+    require 'linguist'
```
Comment (pchaigno) [inline Rakefile]: Why did you have to change this?

**Comment by pchaigno** [comment]
> For a large sample (n=751) of other files with extension .ncl, it is around 92%.

Do you mean 92% files correctly identified as Text?

File: Rakefile (hunk 2 above)
Comment (rpavlick) [inline Rakefile]: `bundle exec rake benchmark:generate` doesn't work for me unless I do that.

**Comment by rpavlick** [comment]
@pchaigno 

Yes, it correctly identifies 92% as Text.

**Comment by arfon** [comment]
> @pchaigno @larsbrinkhoff I decreased the number of added samples.

I'd really like to see this dramatically reduced (see https://github.com/github/linguist/pull/2117). What happens if we reduce to ~5 `NCL` sample files? Does Linguist still do a reasonable job?

**Comment by rpavlick** [comment]
@arfon @pchaigno @larsbrinkhoff 

I reduced it to 16 samples in /samples/NCL and 8 samples in /samples/Text. It correctly classifies my whole set of known NCL files (n=951). I tried with 5 samples but could not get the same level of accuracy.

It incorrectly classifies all of the .ncl files in the github repository fanghuan/ncl. All of these files contain the text strings, "THE_URL:" and "THE_TITLE:". It did not misclassify any other Text files with extension .ncl.

**Comment by pchaigno** [comment]
> I reduced it to 16 samples in /samples/NCL and 8 samples in /samples/Text. It correctly classifies my whole set of known NCL files (n=951).

With 100% accuracy? Does that mean it that the results are better with less samples??

**Comment by rpavlick** [comment]
I also removed a sample from /samples/Text which probably should have been classified as an NCL file. I think that probably made most of the difference, although I didn't test it atomically.

**Comment by pchaigno** [comment]
> I also removed a sample from /samples/Text which should have been classified as an NCL file. I think that probably made most of the difference, although I didn't test it atomically.

Okay, that makes more sense :-)

**Comment by pchaigno** [comment]
I don't get the same results for the Text files :/
```
$ bundle exec linguist ../samples-ncl-text/
79.94%  NCL
20.06%  Text
```

Full results here: https://gist.github.com/pchaigno/2478bc86dd831382bfba

**Comment by rpavlick** [comment]
@pchaigno 

They all look correctly classified to me, except for the ones in 'fanghuan/ncl'. I went manually went through all of the ones where I didn't already know based on the filename or the repository owner.

**Comment by rpavlick** [comment]
It now correctly classifies all of my test files.

File: Rakefile (hunk 2 above)
Comment (pchaigno) [inline Rakefile]: Hum, same for me :/
@arfon Does it work for you?

Comment (arfon) [inline Rakefile]: `require linguist/language` doesn't work for me either :grin:

We restructured some of the requires a while ago and I think this is just out-of-date code.

**Comment by arfon** [comment]
@rpavlick thanks for this. These changes will be live in the next release of Linguist (next couple of weeks)

**Comment by rpavlick** [comment]
Great! Thanks.

**Pull request merged**